### PR TITLE
image: avoid undo multiple Geotag and Datetime signals when not useful

### PR DESCRIPTION
**Pull request by phweyland**
fix slow execution on undo geotag

for 250 images the gpx apply lasts less than 1 sec but undo lasts more than 30 sec
with that fix undo is done in less than 1 sec too.

A question however. Images of group are included in the geotag process. Are the grouped images always at the same location ? (we could want to group images for another reason, couldn't we ?).

### CODE REVIEW

**Review by TurboGit — COMMENTED** [review]
I'm not sure about the implementation. I was expecting to have a GList into dt_undo_geotag_t instead of an imgid to be able to register an offset for multiple images. But maybe I did not understand the goal. Which signal was raised too many time ? Can you comment a bit about what was achieved here? Sorry just a quick look tonight, I'll try to come back tomorrow to it. TIA.

**Comment by phweyland** [comment]
> Can you comment a bit about what was achieved here?

Sure.
Facts.
Testing geotagging on map, I've found this issue about undo of "apply gpx file". In my case I had 266 images geotagged this way. When you "apply gpx file" the control job calculates the new location (specific for each image), applies this location to the image, and this for all the selected images matching the gpx file (266). Only then it emits a signal geotag change (only one).
But undoing this (ctrl-z) freezes the screen for 35 seconds.

Analyze.
A single ctrl-z is enough to undo the 266 images, but the undo routine emits the signal geotag change for each image.
With the current design you can assign a same location to a list of images. In this case undo routine emits the signal geotag change only once.
But the same is not possible in case of a location for each image.

Solution
Instead of setting a list of undo_records (one per image) with is its own before and always the same after,
we set a list of of undo_records (one per image) with is own before and its own after.
In other words, instead of have the parameters: list of images and unique location, the new routine receives the list of images and the corresponding list of locations.

The unique new thing here is to use a g_array for the list of locations instead of a g_list (with a memory allocation for each location).

> I was expecting to have a GList into dt_undo_geotag_t instead of an imgid to be able to register an offset for multiple images

Yes, this was the first idea. It would work for an offset on datetime, but it is not applicable to geotag ...
Here the solution is really generic and could be applied for datetime as well (or any other parameter) if we witness any issue setting the offset. It seems that the consequences for signal image info changed are lighter than for signal geotag change. I have to pursue the tests to check this part.

I hope that clarifies.

Idea: the control job issues a control_log message after having set the geotags (the datetimes as well) with the number of images updated.
Do you think it would be a good idea to do the same in the undo routine ? Or a toast message ?

**Comment by phweyland** [comment]
Pascal,
I'm reviewing the offset/datetime things. This a bit of mess there.
First unlike geotag, offset/datetime doesn't consider the grouped images.
Then timezone stuff has nothing to do (at least today) in image.c but should stay on control_job/geotagging side.
I'm re-coding this part to be make a similar job on both sides, geotag and offset/datetime.

**Comment by TurboGit** [comment]
Ok, just ping me when ready then.

**Comment by phweyland** [comment]
Pascal, that's done.

File: src/common/image.c [hunk 1]
```diff
@@ -502,15 +503,16 @@ void _pop_undo(gpointer user_data, const dt_undo_type_t type, dt_undo_data_t dat
 
       *imgs = g_list_prepend(*imgs, GINT_TO_POINTER(undogeotag->imgid));
       list = g_list_next(list);
+      i++;
     }
-
+    if(i > 1) dt_control_log((action == DT_ACTION_UNDO)
+                              ? _("undone geo location onto %d images")
```
Comment (TurboGit) [inline src/common/image.c:509]: I would reword this as : "geo location undone for %d images"

As undone is a past participle I would put after the subject.

But we may want a review from a native English speaker.

(same below for redone)

File: src/common/image.c [hunk 2]
```diff
@@ -520,9 +522,13 @@ void _pop_undo(gpointer user_data, const dt_undo_type_t type, dt_undo_data_t dat
 
       *imgs = g_list_prepend(*imgs, GINT_TO_POINTER(undodatetime->imgid));
       list = g_list_next(list);
+      i++;
     }
+    if(i > 1) dt_control_log((action == DT_ACTION_UNDO)
+                              ? _("undone date time onto %d images")
```
Comment (TurboGit) [inline src/common/image.c:528]: likewise

**Review by TurboGit — COMMENTED** [review]
Not tested yet.

File: src/common/image.c (hunk 1 above)
Comment (TurboGit) [reply src/common/image.c:509]: @elstoc : What do you think ? TIA.

Comment (phweyland) [reply src/common/image.c:509]: I've just adapted the control-job messages ("added time offset to ...", "set date time to...", "applied matched GPX location ...".
They may have to be revisited too ...

Comment (elstoc) [reply src/common/image.c:509]: @TurboGit I agree "geo location undone for %d images" is better. I've amended references to "date time" in the user manual to read "date/time" and I think they read better this way so, while you're changing it best to be consistent. The only reference I've left unchanged in dtdocs is the module control itself, so that'll need to be updated if you choose to change it here.

"redone" would probably read better as "re-applied to"

Comment (phweyland) [reply src/common/image.c]: done!

Comment (phweyland) [reply src/common/image.c]: @elstoc, I've changed all the date time to date/time here. 
I've a commit ready to align the date/time wording in dtdocs.

Comment (elstoc) [reply src/common/image.c]: Thanks @phweyland

Comment (elstoc) [reply src/common/image.c]: Another minor thing while we're discussing wording - I think "geo location" might read better if it's hyphenated "geo-location". No need to necessarily do that in this PR but something for consideration later (since it will possibly have implications elsewhere). I suppose more generally it's arguable whether the "geo" is required at all.

**Comment by elstoc** [comment]
(I've moved this comment so it shows in the main thread)

Another minor thing while we're discussing wording - I think "geo location" might read better if it's hyphenated "geo-location". No need to necessarily do that in this PR but something for consideration later (since it will possibly have implications elsewhere). I suppose more generally it's arguable whether the "geo" is required at all.

**Comment by phweyland** [comment]
> I think "geo location" might read better if it's hyphenated "geo-location".

Should be easy. There are few occurrence of "geo location".

> I suppose more generally it's arguable whether the "geo" is required at all.

Without being mandatory "geo" is consistent with "geotagging". On the other hand, "location" alone may not be enough to describe gps coordinates...

**Comment by elstoc** [comment]
> Without being mandatory "geo" is consistent with "geotagging". On the other hand, "location" alone may not be enough to describe gps coordinates...

Yeah I was less sure about that one. Fine to leave as is.

File: src/control/jobs/control_jobs.c [hunk 3]
```diff
@@ -1872,33 +1918,52 @@ static int32_t dt_control_datetime_job_run(dt_job_t *job)
 
   const guint total = g_list_length(t);
 
-  const char *mes11 = offset ? N_("adding time offset to %d image") : N_("setting date time to %d image");
-  const char *mes12 = offset ? N_("adding time offset to %d images") : N_("setting date time to %d images");
+  const char *mes11 = offset ? N_("adding time offset to %d image") : N_("setting date/time to %d image");
```
Comment (elstoc) [inline src/control/jobs/control_jobs.c:1921]: This should read "setting date/time **of** %d image". The "adding time offset to..." is fine. Similarly below

Comment (phweyland) [reply src/control/jobs/control_jobs.c:1921]: Can I disagree here ? :smile: 
Adding `time offset to %d image` is less correct than `setting date/time to %d image`.
Time offset should added to date/time of an image and not to an image.
If **of** is OK, `setting date/time to %d image` doesn't sound really wrong (the image gets a date/time).

Comment (elstoc) [reply src/control/jobs/control_jobs.c:1921]: You can disagree if you like but "setting date/time **to** 1 image" is wrong. If you want to say "setting time/date of 1 image to DD/MM/YYYY" that's fine.

Comment (elstoc) [reply src/control/jobs/control_jobs.c:1921]: you could say "applying date/time to %d image" and that would be ok

Comment (phweyland) [reply src/control/jobs/control_jobs.c:1921]: > Don't ask me to explain why, 

I won't, every language has its own secrets. :smile:
I'll change to **of**.

**Comment by phweyland** [comment]
Wording changes requested done.

**Review by elstoc — APPROVED** [review]
